### DISCUSSION TITLE
loop during AP

### DISCUSSION THREAD

**Discussion by jeppo7712** [HomeSpan/HomeSpan#651] [Ideas]
When no Wifi is configured, the AP is started, and the Arduino loop is blocked. I might be a good idea to either make the AP loop not blocking, or add the possibility to call a user defined function inside the AP loop so that critical functions during the AP mode can be called...

**Comment by HomeSpan** [comment]
The AP mode is designed for setup only, so should not be called during normal operation.   Also, during AP mode your phone is not connected to WiFi, which limits what you can do with HomeSpan.   Can you give some examples of the type of operations would you want to perform while the AP mode is running?   It may be possible to add a loop function in the background in a future release.

**Comment by jeppo7712** [comment]
For example, a water leak detector which should switch a relay as soon as a water leak is detected. This is my use case actually: I would like that this works as soon as I power the device, independently of the wifi status -even if wifi is not configured due to whatever reason, and homespan starts the AP. The sensor code is called in the main loop, but it could be blocked by homespan.poll... 

Using autopoll might trigger race conditions because my leak sensor and relay switch code is talking to homespan's Leak Sensor and Valve homekit accessories, which would then be running on an separate CPU task... (I don't want to start using task queues etc...).

So a loop call in AP mode would solve some of these issues.

I'm loving homespan btw... ;)

**Comment by HomeSpan** [comment]
Thanks - this is a really good use case.  I can very likely add a callback in the AP function to run a user-defined function as you described.  However, I'm wondering if in addition to this it would make sense for the AP function to automatically call the loop() commands of each service every cycle.  This would allow you to integrate the whatever logic is needed directly in the service's loop() function, and it would automatically be called from the AP as well.

**Comment by jeppo7712** [comment]
great! good to hear! Calling the service loops would be even more awesome ;)

**Comment by jeppo7712** [comment]
Continuing with the same topic (critical functions during AP), does Spanpoint work during AP mode? I would like to use a remote water sensor, which could send data using Spanpoint as soon as a leak is detected, and like before, it shouldn't matter in which state the wifi is for this critical function...
